refactor: cleanup on public API

tlambert03 opened the pull request:
This renames "controller" to `ArrayViewer`, and cleans up a number of naming and API issues in preparation for moving the v2-mvc branch into `main`.

- moves all of the `old_viewer` stuff into `v1` module
- all examples now use v2
- `ViewerController` renamed to `ArrayViewer` and API cleaned up.  
- axis normalization logic is moved to the `DataWrapper`.
- Many of the v1 datawrapper subclasses are ported into v2 (e.g. sparse, pyopencl, etc...)
- `imshow` is updated to use v2
- misc other small changes that fix various issues that popped up as as made these changes
 
-----

~I imagine the distinction between `ArrayDataDisplayModel` and `ArrayDisplayModel` is probably still a little unclear to an end-user, and will require some documentation.    In practice, they will mostly only need to care about `ArrayDataDisplayModel` (but I suspect some understanding of both is still necessary)~

_edit:_. I left the `ArrayDataDisplayModel` as a private implementation detail for now.  the `ArrayDisplayModel` and `DataWrapper` are the main things exposed publicly

tlambert03 commented:
@gselzer ... this is ready for your eyes if you'd like to have a look.  After this merges, I intend to merge v2-mvc with `main` and make it the default branch again.  and probably start releasing `rc` versions on PyPI.

it's a medium/large PR (sorry) but it's mostly name changes and method re-arrangement

gselzer commented on tests/test_mvc_viewer.py:
Or maybe it could be merged with `test_controller.py`?

tlambert03 replied:
like that, done

gselzer commented on tests/test_examples.py:
```diff
@@ -9,7 +9,9 @@
 
     if pytestqt.qt_compat.qt_api.pytest_qt_api.startswith("pyside"):
-        pytest.skip("V1 viewer segfaults with pyside", allow_module_level=True)
+        pytest.skip(
+            "viewer still occasionally segfaults with pyside", allow_module_level=True
+        )
```
😆

gselzer commented on src/ndv/views/_vispy/_vispy.py:
I know this is internal code, but since we're moving things around...what do you think about a rename to `_array.py`? And same for `src/views/_pygfx/_pygfx.py`?

tlambert03 replied:
sure!

gselzer reviewed (commented):
This looks great!

I'm still trying to get my head around `ArrayDataDisplayModel`...I guess I still don't really understand the need. Do you think it's possible to minimize or even fully avoid the exposure of that class? I think that, especially with `ArrayDisplayModel` floating around with a *very* similar name, there will be lots of confusion if accessible to users...

What's your plan for v1? Keep it around through the pre-releases?

tlambert03 commented:
> I'm still trying to get my head around ArrayDataDisplayModel...I guess I still don't really understand the need.

did you read the docstring [here](https://github.com/pyapp-kit/ndv/pull/80/files#diff-3fa91e1826b0fbd9e536801287147a39ec35b9a68e1be288048018dd9b8b5714R39-R59) and still not find it helpful?

The key need here is described here:

>    Having this class composed of the two other models (rather than inheriting from
     `ArrayDisplayModel`) allows for multiple models to share the same underlying
     display model (e.g. for linked views).

`ArrayDisplayModel` is a "prescriptive" display model, that may use abstract labels in the absence of data.  It can say "visible axes are `X`, `Y`, and `Z`" ... 


`ArrayDataDisplayModel` *has* (not *is*) a display model, and resolves the prescription for a specific dataset.  This allows for a many to many relationship:


```python
display_model_1 = ArrayDisplayModel()
display_model_2 = ArrayDisplayModel()

data1 = DataWrapper()
data2 = DataWrapper()

# show the SAME data with different display, such might be done in an OrthoViewer
model1 = ArrayDataDisplayModel(data1, display_model1)
model2 = ArrayDataDisplayModel(data1, display_model2)

# show the SAME display model with different datasets, such as might be done to
# compare two different processing results or a synchronized view of two different things
model3 = ArrayDataDisplayModel(data1, display_model1)
model4 = ArrayDataDisplayModel(data2, display_model1)
```

At the end of the day, most end users will only see the combined `ArrayDataDisplayModel`.  It's what you get when you access `viewer.model` (*not* the `ArrayDisplayModel`).  So while `ArrayDisplayModel` was the first one we discussed internally, it's not being presented as first thing you see publicly here.  We associate it with data first. 

if the whole many-to-many relation makes sense, then I wonder if it's just the similarity of the names that is confusing?  would you propose a different name?  or a different construction?

tlambert03 commented on src/ndv/controllers/_array_viewer.py:
```diff
@@ -1,37 +1,78 @@
-    def __init__(self, data: DataDisplayModel | None = None) -> None:
+    def __init__(
+        self,
+        data_or_model: Any | DataWrapper | ArrayDataDisplayModel = None,
```
> I feel like the extra complication of the ArrayDataDisplayModel should be internalized if it can be...

it's the thing you get back from `.model()` (*not* ArrayDisplayModel) ... and i expose it because otherwise the end user has no easy way to ask "hey what axis *is* the `"Z"` or `-3` or `"generic-label"` anyway?).  It passes through all the getters and setters of the wrapped `ArrayDisplayModel` (so one doesn't need to access that wrapped object unless they want to do something interesting with linked views, as described in my other comment)

tlambert03 replied:
That said, I don't mind removing it from `__init__`... however, if someone wanted to duplicate the viewer, then instead of `

```python
viewer2 = ArrayViewer(viewer1.model)
```

they would need to  do

```python
viewer2 = ArrayViewer(viewer1.model.data_wrapper, viewer1.model.display_model)
```

gselzer replied:
Oh, yeah, duplicating the viewer could be nice. Some nice convenience methods could be:
```python
viewer2 = ArrayViewer(viewer1)
```

or
```python
viewer2 = viewer1.clone()
```

gselzer replied:
> it's the thing you get back from `.model()` (_not_ ArrayDisplayModel) ... and i expose it because otherwise the end user has no easy way to ask "hey what axis _is_ the `"Z"` or `-3` or `"generic-label"` anyway?). It passes through all the getters and setters of the wrapped `ArrayDisplayModel` (so one doesn't need to access that wrapped object unless they want to do something interesting with linked views, as described in my other comment)

Doesn't this PR move that functionality into the `DataWrapper? Can the end user could just ask the data wrapper what that is?

gselzer commented:
> if the whole many-to-many relation makes sense, then I wonder if it's just the similarity of the names that is confusing? would you propose a different name? or a different construction?

Oh, yeah, I absolutely understand the need for the many-to-many relation - I'm just wondering whether we can do without the `ArrayDataDisplayModel` class. It's an extra layer of abstraction that users will need to understand, which I'd prefer to avoid.

As long as we're pre-releasing, let's not worry about it. Maybe I can take some time before a real release to understand why it's necessary and/or whether it could be removed.

tlambert03 commented:
>  I'm just wondering whether we can do without the ArrayDataDisplayModel class. It's an extra layer of abstraction that users will need to understand, which I'd prefer to avoid.

how would you construct a many to many relationship between display and data without some addition layer object?

```python
class Joiner:
    display: object
    model: object
```

tlambert03 commented:
ah... maybe you're not arguing the need for the object, but rather whether it has to be exposed? (i know you're literally saying that, and I'm dense :joy:).  I mentioned in https://github.com/pyapp-kit/ndv/pull/80#discussion_r1907753996 why I think i wanted to expose it.  (because it's a useful object that resolves axis keys, and it contains all the slicing logic).  but we could perhaps hide it for now and expose it in a less prominent way only if need be

gselzer commented:
> > I'm just wondering whether we can do without the ArrayDataDisplayModel class. It's an extra layer of abstraction that users will need to understand, which I'd prefer to avoid.
> 
> how would you construct a many to many relationship between display and data without some addition layer object?
> 
> ```python
> class JoinModel:
>     display: object
>     data: object
> ```

well, the current situation is more or less:

```python
class ArrayModel:
    display: DisplayModel
    data: DataModel

class ArrayViewer:
    model: ArrayModel
```

I'm wondering whether it could be:

```python
class ArrayViewer:
    display: DisplayModel
    data: DataModel
```

I'm guessing that since you made such a pretty `ArrayModel` class, there's some reason why we need a class to encapsulate the m2m relationship. I just don't understand that reason (yet).

> ah... maybe you're not arguing the need for the object, but rather whether it has to be exposed?

Well, I'm wondering about the former, but if we don't expose it then I don't have to wonder **really hard** *right now* 😅. I would be thrilled if we could find a way to not expose it, because that would contain the confusion we mentioned earlier to just the devs.

tlambert03 commented:
> Well, I'm wondering about the former, but if we don't expose it then I don't have to wonder really hard right now 😅

lol.  yes this is indeed all very revealing :) it's probably a bad design :joy:

> I'm guessing that since you made such a pretty ArrayModel class, there's some reason why we need a class to encapsulate the m2m relationship. I just don't understand that reason (yet).

gotcha!  K, the reason I made that class (and it might *not* be necessary) was primarily as a place to 

1. store/cache normalized axis keys when an `ArrayDisplayModel` is joined with a specific `DataWrapper`... this is all of those `normalized_current_index` and other similar properties.
2. use the above methods to actually generate request/response objects that slice and return the data.

However, I ended up moving part of `1.` into the `DataWrapper` itself.  so we could definitely hide those additional `normalized_x` methods somewhere inside of the controller.  and the methods in `2` really don't  *need* to be on a class either.

sooooo.... lemme branch off this and try to get rid of it again and see how it feels!

tlambert03 commented:
(thank you by the way for your patience :)

gselzer commented:
> However, I ended up moving part of `1.` into the `DataWrapper` itself. so we could definitely hide those additional `normalized_x` methods somewhere inside of the controller. and the methods in `2` really don't _need_ to be on a class either.
> 
> sooooo.... lemme branch off this and try to get rid of it again and see how it feels!

Well, part of the reason that it took me so long to get through this was because I was experimenting with moving both 1. and 2. within the `DataWrapper` class. Then, data requests method(s) on the `DataWrapper` could either accept the `ArrayDisplayModel` directly, or could accept parts of it. I paused this, though, because I wanted to convey my general approval through code review - I'm still not sure whether a design like that would be better than what you have now, though.

> (thank you by the way for your patience :)

And thank you for being patient with me!

tlambert03 commented:
ok, in the [last two commits](https://github.com/tlambert03/ndv/compare/3ee6d83...tlambert03:ndv:77a5547) I removed public usage/mention of the `ArrayDataDisplayModel`.  I retain the class internally as a convenience class to organize some of the functionality, but i removed all of the pass-through convenience `@property` access from it, and gave it a private name.  (I don't want to move the slicing to `DataWrapper` just yet, in this PR, but not opposed to looking into that later).

The `ArrayViewer` now has public access to the `DataWrapper` and the `ArrayDisplayModel`.

let me know what you think @gselzer

gselzer commented on src/ndv/controllers/_array_viewer.py:
```diff
@@ -1,37 +1,63 @@
-    This is the primary public interface for the viewer.
+    Parameters
+    ----------
+    data_or_model :  DataWrapper | Any
```
```suggestion
    data :  DataWrapper | Any
```

tlambert03 replied:
good catch

tlambert03 replied:
still waiting on https://github.com/astral-sh/ruff/issues/924 :)

tlambert03 commented on tests/test_controller.py:
```diff
@@ -144,10 +148,35 @@ def test_histogram_controller() -> None:
     mock_histogram.set_data.reset_mock()
-    ctrl.model.current_index.assign({0: 1, 1: 2, 3: 3})
+    ctrl.display_model.current_index.assign({0: 1, 1: 2, 3: 3})
     mock_histogram.set_data.assert_called_once()
 
     # switching to composite mode puts the histogram view in the
     # lut controller for all channels (this may change)
-    ctrl.model.channel_mode = ChannelMode.COMPOSITE
+    ctrl.display_model.channel_mode = ChannelMode.COMPOSITE
```
sure :)

tlambert03 replied:
... but, then would it be confusing to have `.view` and `.display`?

gselzer replied:
Oh yeah, good point 🤔 

Let's just leave it for now

gselzer commented:
> ok, in the [last two commits](https://github.com/tlambert03/ndv/compare/3ee6d83...tlambert03:ndv:77a5547) I removed public usage/mention of the `ArrayDataDisplayModel`. I retain the class internally as a convenience class to organize some of the functionality, but i removed all of the pass-through convenience `@property` access from it, and gave it a private name. (I don't want to move the slicing to `DataWrapper` just yet, in this PR, but not opposed to looking into that later).

I love it, thank you so much @tlambert03. Do you think we lost any functionality in the process?

Looking once more at the public `ArrayViewer` API, I'm pretty dang happy with how simple the API is:

![image](https://github.com/user-attachments/assets/868540cf-9c14-43fe-863c-66f33f696e8a)

Couple final thoughts:
* `show` and `hide` could become a single `visible` property.
* Do you think it makes sense to make `add_histogram` private? The API for histograms will likely change later as we better support multiple histograms for multiple channels...just a thought that I'm not tied to, and you do test that method so 🤷

tlambert03 commented:
> Do you think we lost any functionality in the process?

not at all...  i like it better for now as well

>  I'm pretty dang happy with how simple the API is:

me too 👍

> Do you think it makes sense to make add_histogram private?

I do.  will do.

> Is there a reason for exposing the view? Isn't the idea that users shouldn't be interacting with that?

Well, it would be nearly impossible to use this as a component in some broader context (such as in a pymmcore-gui) without having some way to access the view widget.  so we need to expose *something*.  It could just be the `frontend_widget`... and that already opens the possibility that a user will do something problematic.  But I think the alternative is simply too restrictive, to the degree that this really just becomes an end-point application and not a component that can be reused.   So i think it's a risk that has to be taken (along with the warning i added in the docs).  whether we expose the full `view` or just `frontend_widget` is a reasonable question.  but i think let's just start with view for now and deal with deprecation if we actually hit problems with that

tlambert03 commented:
> show and hide could become a single visible property.

yeah... but `.show()` is *such* a common API for pretty much every gui.  So I feel a little "required" here to do the standard thing.  that could be in addition to `.visible = ` ... but i kinda dislike property-setters on the whole.  so let's punt on that discussion
